Feature/st4table

mickaelaccensi opened the pull request:
# Pull Request Summary
remove lookup table for ST4 and clean up the ST4 code

## Description
remove lookup table in ST4 to speed up computation
clean up ST4 code 
correct uninitialized variables

### Issue(s) addressed
- fixes #1085 

### Commit Message
remove lookup table for ST4 to speed up computation and clean up the ST4 code

### Check list  

- [X] Branch is up to date with the authoritative repository (NOAA-EMC) develop branch. 
- [X] Checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop). 
- [?] If a version number update is required, checked the [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) checklist. 
- [X] If a new feature was added, a regression test for testing the new feature is added. 
 
### Testing

* How were these changes tested? matrix
* Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) ww3_ts1
* Have the matrix regression tests been run (if yes, please note HPC and compiler)? yes
* Please indicate the expected changes in the regression test output, (Note the [list](https://github.com/NOAA-EMC/WW3/wiki/How-to-use-matrix.comp-to-compare-regtests-with-develop#4-look-at-results) of known non-identical tests.) mod_def.ww3, out_grd.ww3, out_pnt.ww3, restart.ww3 for all runs using ST4 switch
* Please provide the summary output of matrix.comp (_matrix.Diff.txt_, _matrixCompFull.txt_ and _matrixCompSummary.txt_) 
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/13873741/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/13873744/matrixCompSummary.txt)
matrixDiff.txt is too big to be uploaded

**expected changes :**
ww3_ts1 : output parameters slight changes (work_T7**)
ww3_ts2 : only mod_def since last bugfix
ww3_ts3 : only mod_def since last bugfix
ww3_ts4 : only mod_def
ww3_tp2.6 : only mod_def
ww3_tp2.6/work_pdlib : not b4b due to scotch library on Ifremer HPC (also on develop branch)
ww3_tp2.18 : only mod_def
ww3_tp2.14 : only mod_def since last bugfix
ww3_tic2.3 : only mod_def
ww3_tic1.4 : only mod_def
ww3_ta1 :  only mod_def since last bugfix
mww3_test_08 : only mod_def since last bugfix
mww3_test_05 : only mod_def since last bugfix
mww3_test_03 : known not b4b
ww3_tp1.8 : output parameters changes for taw, two and foc (partially correct issue in T570)

**could be expected differences :**
ww3_tp2.21 : output parameters slight changes (work_b and work_mb)
ww3_tp2.17 : output parameters slight changes (work_c, work_mc, work_mc1)
ww3_tp2.15 : output parameters slight changes

mickaelaccensi committed:
start to integrate ST4 updates

mickaelaccensi committed:
import ST4 changes

mickaelaccensi committed:
add regtest

mickaelaccensi committed:
Merge branch 'develop' into feature/ST4table

mickaelaccensi committed:
add SINTAILPAR init

mickaelaccensi committed:
Merge branch 'develop' into feature/ST4table

mickaelaccensi committed:
Merge branch 'develop' into feature/ST4table

mickaelaccensi committed:
add namelists

mickaelaccensi committed:
address issue #1085
address issue #1074
add regtest

mickaelaccensi committed:
Merge branch 'develop' into feature/ST4table

mickaelaccensi committed:
Merge branch 'develop' into feature/ST4table

mickaelaccensi committed:
correct typo

mickaelaccensi committed:
remove comment

mickaelaccensi committed:
correct typo

mickaelaccensi committed:
correct bug in mod_def reading

mickaelaccensi committed:
add variable initialization
correct duplicated variable read/write

MatthewMasarik-NOAA commented:
@mickaelaccensi, this PR is up next, I'll put the regression tests in now!

MatthewMasarik-NOAA commented:
Hi @mickaelaccensi, the header mentions the matrix.comp tests were running when this was submitted.  Have those completed?

mickaelaccensi commented:
completed but I'm not really happy with the differences... they are many regtests with different results so I have to discuss with Fabrice first. I'll keep informed asap

MatthewMasarik-NOAA commented:
Hi @mickaelaccensi, okay, thank you for the update.  Yes, please keep us informed when you find out more.

JessicaMeixner-NOAA commented on model/src/ww3_ounp.F90:
```diff
@@ -2158,7 +2158,7 @@ SUBROUTINE W3EXNC(I,NCID,NREQ,INDREQ,ORDER)
 #endif
             !
-            DO ITT=1, 3
+            DO ITT=1, 4
```
@mickaelaccensi I was looking through these updates and saw this one line change that seemed a bit out of place given everything else.  My guess is  this likely isn't the issue causing the unexpected diffs, but wanted to point it out just in case.

mickaelaccensi committed:
manually merge bgufix/gqm

MatthewMasarik-NOAA commented:
@mickaelaccensi, with #1127 merged I'm going to start in on this PR.  Do you have the matrix logs?  This is a big PR with a number of expected differences, though I'm going to try to keep it moving along as Chris needs it merged for his work.

mickaelaccensi commented:
@MatthewMasarik-NOAA  I first have to solve the issues found in regtests mww3_test_05, ww3_ts2 and ww3_ts3 which contains too large Hs values. I'll work on it this week and keep you informed

MatthewMasarik-NOAA commented:
Hi @mickaelaccensi, okay I understand, thanks for the update.  I'll wait to hear from you to proceed.

Commit:
Corrected type of VISCSTRESS from INTEGER to REAL

mickaelaccensi committed:
Merge branch 'feature/ST4table' of https://gitlab.ifremer.fr/wave/WW3 into feature/ST4table

mickaelaccensi committed:
Merge branch 'develop' into feature/ST4table

mickaelaccensi committed:
correct typo

Commit:
Fixed bug with undefined LEVTAIL used instead of LEVTAIL0 when table is
used in W3SIN4

Commit:
merging before pushing LEVTAIL bug correction
Merge branch 'feature/ST4table' of https://gitlab.ifremer.fr/wave/WW3 into feature/ST4table

mickaelaccensi committed:
Merge branch 'develop' into feature/ST4table

mickaelaccensi commented:
finally ! the branch is ready to be merged ! 
sorry for this long delay @ukmo-ccbunney

JessicaMeixner-NOAA commented:
@mickaelaccensi - I'll be doing the testing for this PR.  Our main machine we test on is down for maintenance today, but I will submit tests first thing tomorrow.

MatthewMasarik-NOAA commented:
Thanks for the updates @mickaelaccensi!  I will be going on leave so @JessicaMeixner-NOAA is kindly  taking care of this review.

JessicaMeixner-NOAA commented:
@mickaelaccensi my list of not b4b tests is mostly the same as yours with a quick glance.  I have a few more diffs in the ufs tests that we run and you do not and so I want to look at that. Also, since there's quite a few differences, I'm going to do a sanity test that we can reproduce answers with this PR.  Our queues are pretty slow right now and despite submitting the runs last night, they have not yet started running, so this is unfortunately going to take longer than I'd hope.  I'll keep you posted if I run into anything unexpected or have questions.

JessicaMeixner-NOAA commented on model/src/w3src4md.F90:
```diff
@@ -294,17 +293,18 @@ SUBROUTINE W3SPR4 (A, CG, WN, EMEAN, FMEAN, FMEAN1, WNMEAN,     &
     DO IK=1, NK
       EB(IK)  = 0.
       EB2(IK) = 0.
+      SIGFAC=SIG(IK)**SSDSC(12) * DDEN(IK) / CG(IK)
```
There's never a scenario where CG(IK) =0 is there?  (I'm assuming no, but if so we should have some sort of catch so we do not divide by zero).

JessicaMeixner-NOAA commented:
@mickaelaccensi the queue times are really long right now, so unfortunately I'm still in the queue waiting.  In the meantime, I went back through the regtests and things look as expected: 

```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR3_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (12 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (10 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (16 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (16 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (12 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (16 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e_c                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (15 files differ)
mww3_test_05/./work_ST4_PR3_UQ_MPI_OMPH                     (10 files differ)
mww3_test_05/./work_ST4_PR3_UNO_MPI                     (10 files differ)
mww3_test_05/./work_ST4_PR3_UQ_MPI                     (10 files differ)
mww3_test_05/./work_ST4_PR2_UQ_MPI                     (10 files differ)
mww3_test_05/./work_ST4_PR2_UNO_OMP                     (10 files differ)
mww3_test_05/./work_ST4_PR3_UNO_OMP                     (10 files differ)
mww3_test_05/./work_ST4_PR3_UNO_MPI_OMPH                     (10 files differ)
mww3_test_05/./work_ST4_PR1_MPI                     (10 files differ)
mww3_test_05/./work_ST4_PR1_MPI_OMPH                     (10 files differ)
mww3_test_05/./work_ST4_PR1_OMP                     (10 files differ)
mww3_test_05/./work_ST4_PR3_UQ_OMP                     (10 files differ)
mww3_test_05/./work_ST4_PR2_UNO_MPI                     (10 files differ)
mww3_test_05/./work_ST4_PR2_UNO_MPI_OMPH                     (10 files differ)
mww3_test_05/./work_ST4_PR2_UQ_OMP                     (10 files differ)
mww3_test_05/./work_ST4_PR2_UQ_MPI_OMPH                     (10 files differ)
mww3_test_08/./work_ST4_PR3_UQ_MPI                     (4 files differ)
mww3_test_08/./work_lowres                     (2 files differ)
mww3_test_09/./work_MPI_ASCII                     (0 files differ)
ww3_ta1/./work_UPD0F_O                     (2 files differ)
ww3_ta1/./work_UPD0F_U                     (2 files differ)
ww3_ta1/./work_UPD2_U_cap                     (2 files differ)
ww3_ta1/./work_UPD3_U                     (2 files differ)
ww3_ta1/./work_UPD5_U_cap                     (2 files differ)
ww3_ta1/./work_UPD6_O                     (2 files differ)
ww3_ta1/./work_UPD2_U                     (2 files differ)
ww3_ta1/./work_UPD5_O                     (2 files differ)
ww3_ta1/./work_UPD5_U                     (2 files differ)
ww3_ta1/./work_UPD3_U_cap                     (2 files differ)
ww3_ta1/./work_UPD3_O                     (2 files differ)
ww3_ta1/./work_UPD6_U                     (2 files differ)
ww3_ta1/./work_UPD2_O                     (2 files differ)
ww3_ta1/./work_UPD6_U_cap                     (2 files differ)
ww3_tic1.4/./work_IC1IS2_1000                     (1 files differ)
ww3_tic1.4/./work_IC2IS2_IC2d                     (1 files differ)
ww3_tic1.4/./work_IC2IS2_IC2b                     (1 files differ)
ww3_tic1.4/./work_IC0IS2_1000                     (1 files differ)
ww3_tic2.3/./work_IC2IS2creep                     (1 files differ)
ww3_tic2.3/./work_IC2IS2scat                     (1 files differ)
ww3_tic2.3/./work_IC2IS2dissip                     (1 files differ)
ww3_tp1.8/./work_PR3_UQ                     (6 files differ)
ww3_tp2.10/./work_MPI                     (2 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (9 files differ)
ww3_tp2.14/./work_OASACM6                     (2 files differ)
ww3_tp2.14/./work_OASACM3                     (2 files differ)
ww3_tp2.14/./work_OASACM5                     (2 files differ)
ww3_tp2.14/./work_OASACM4                     (2 files differ)
ww3_tp2.14/./work_OASACM                     (2 files differ)
ww3_tp2.14/./work_OASICM                     (2 files differ)
ww3_tp2.14/./work_OASACM2                     (2 files differ)
ww3_tp2.14/./work_OASOCM                     (2 files differ)
ww3_tp2.15/./work_PR3_UQ_RHO                     (7 files differ)
ww3_tp2.15/./work_5km                     (2 files differ)
ww3_tp2.15/./work_PR3_UQ_RHO_MPI                     (7 files differ)
ww3_tp2.15/./work_PR3_UQ                     (2 files differ)
ww3_tp2.15/./work_MPI_5km                     (2 files differ)
ww3_tp2.15/./work_PR3_UQ_MPI                     (2 files differ)
ww3_tp2.15/./work_ST4FLX5                     (7 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (4 files differ)
ww3_tp2.17/./work_ma                     (2 files differ)
ww3_tp2.17/./work_a                     (2 files differ)
ww3_tp2.17/./work_mc1                     (2 files differ)
ww3_tp2.17/./work_mb                     (2 files differ)
ww3_tp2.17/./work_mc                     (2 files differ)
ww3_tp2.17/./work_ma1                     (2 files differ)
ww3_tp2.17/./work_c                     (2 files differ)
ww3_tp2.17/./work_b                     (2 files differ)
ww3_tp2.18/./work_TIDE_MPI                     (2 files differ)
ww3_tp2.18/./work_TIDE                     (2 files differ)
ww3_tp2.21/./work_ma                     (4 files differ)
ww3_tp2.21/./work_b_metis                     (2 files differ)
ww3_tp2.21/./work_a                     (2 files differ)
ww3_tp2.21/./work_mb                     (4 files differ)
ww3_tp2.21/./work_b                     (2 files differ)
ww3_tp2.6/./work_ST4                     (2 files differ)
ww3_tp2.6/./work_ST4_ASCII                     (3 files differ)
ww3_ts1/./work_Romero                     (directory not found)
ww3_ts1/./work_T701                     (directory not found)
ww3_ts1/./work_ST4_WRT                     (2 files differ)
ww3_ts1/./work_ST4_GMD                     (2 files differ)
ww3_ts1/./work_ST4                     (2 files differ)
ww3_ts1/./work_ST4_TSA                     (2 files differ)
ww3_ts1/./work_ST4_T700                     (6 files differ)
ww3_ts1/./work_T702                     (directory not found)
ww3_ts1/./work_ST4_T500                     (directory not found)
ww3_ts1/./work_T707GQM                     (6 files differ)
ww3_ts1/./work_T713GQM                     (6 files differ)
ww3_ts2/./work_ST4_PR1                     (2 files differ)
ww3_ts2/./work_ST4_PR3_UNO                     (2 files differ)
ww3_ts2/./work_ST4_PR3_UQ                     (2 files differ)
ww3_ts2/./work_ST4_PR2_UQ                     (2 files differ)
ww3_ts2/./work_ST4_PR2_UNO                     (2 files differ)
ww3_ts3/./work_ST4_PR3_UQ_MPI_OMPH                     (2 files differ)
ww3_ts3/./work_ST4_PR3_UNO_MPI                     (2 files differ)
ww3_ts3/./work_ST4_PR3_UQ_MPI                     (2 files differ)
ww3_ts3/./work_ST4_FLD2_MPI                     (2 files differ)
ww3_ts3/./work_ST4_PR2_UQ_MPI                     (2 files differ)
ww3_ts3/./work_ST4_PR2_UNO_OMP                     (2 files differ)
ww3_ts3/./work_ST4_PR3_UNO_OMP                     (2 files differ)
ww3_ts3/./work_ST4_PR3_UNO_MPI_OMPH                     (2 files differ)
ww3_ts3/./work_ST4_PR1_MPI                     (2 files differ)
ww3_ts3/./work_ST4_PR1_MPI_OMPH                     (2 files differ)
ww3_ts3/./work_ST4_FLD1_MPI                     (2 files differ)
ww3_ts3/./work_ST4_PR1_OMP                     (2 files differ)
ww3_ts3/./work_ST4_PR3_UQ_OMP                     (2 files differ)
ww3_ts3/./work_ST4_PR2_UNO_MPI                     (2 files differ)
ww3_ts3/./work_ST4_PR2_UNO_MPI_OMPH                     (2 files differ)
ww3_ts3/./work_ST4_PR2_UQ_OMP                     (2 files differ)
ww3_ts3/./work_ST4_PR2_UQ_MPI_OMPH                     (2 files differ)
ww3_ts4/./work_rg_shel_MPI                     (1 files differ)
ww3_ts4/./work_ug_MPI                     (1 files differ)
ww3_ts4/./work_rg_multi_MPI                     (3 files differ)
ww3_ufs1.1/./work_unstr_b                     (2 files differ)
ww3_ufs1.1/./work_c_npl                     (4 files differ)
ww3_ufs1.1/./work_unstr_a                     (2 files differ)
ww3_ufs1.1/./work_unstr_c                     (2 files differ)
ww3_ufs1.1/./work_d                     (4 files differ)
ww3_ufs1.1/./work_c_nth                     (4 files differ)
ww3_ufs1.1/./work_c                     (4 files differ)
ww3_ufs1.2/./work_a                     (12 files differ)
ww3_ufs1.2/./work_l                     (8 files differ)
ww3_ufs1.2/./work_c                     (14 files differ)
ww3_ufs1.2/./work_b                     (12 files differ)
ww3_ufs1.3/./work_a                     (13 files differ)
```
 
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/13922780/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/13922783/matrixCompSummary.txt)


I also did another check of the code and noted 2 places where we may or may not want to have a catch for divide by zero.   In the meantime, I do want to make sure we can reproduce results before merging so apologies for the delay on that.  Monday is a holiday here in the US, so it'll likely be 1/16 for merge right now.  One other comment: Did we need to add any namelist description updates to model/inp/ww3_grid.inp?

JessicaMeixner-NOAA commented:
@mickaelaccensi thank you for your patience with the slow queues on this end.  I have finally been able to confirm everything is b4b with itself: 

```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR1_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (16 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (11 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (10 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (16 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (18 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (16 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e_c                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (16 files differ)
mww3_test_09/./work_MPI_ASCII                     (0 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (6 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (4 files differ)
ww3_tp2.6/./work_ST4_ASCII                     (0 files differ)
ww3_ufs1.3/./work_a                     (3 files differ)
```

JessicaMeixner-NOAA reviewed (approved):
@mickaelaccensi  I'm approving this PR, however, it would be great if you could follow up on two things still: 

* Confirm the two spots in w3src4md will not ever be divide by zero (a follow-up PR for work around is sufficient if this might ever happen). 
* If anything needs to be added to a namelist description, please submit another PR. 

There are many changes described in PR text, but I have confirmed this branch reproduces itself.

mickaelaccensi commented:
> @mickaelaccensi I'm approving this PR, however, it would be great if you could follow up on two things still:
> 
>     * Confirm the two spots in w3src4md will not ever be divide by zero (a follow-up PR for work around is sufficient if this might ever happen).
It should not happen but I'll confirm that with Fabrice
> 
>     * If anything needs to be added to a namelist description, please submit another PR.
I'm working on it, I just  need some additional information from Fabrice, then I'll make a PR
> 
> 
> There are many changes described in PR text, but I have confirmed this branch reproduces itself.

JessicaMeixner-NOAA commented:
Thanks @mickaelaccensi !

mickaelaccensi commented:
I confirm that division by CG=0 should never happen

JessicaMeixner-NOAA commented:
> I confirm that division by CG=0 should never happen

Thank you!!!
